racstyle/code-challenge-datamade
# Challenge solution

## Overview

This is Rachel's solution to DataMade's code challenge for their Developer role.  It is a recreation of DataMade's [Parserator](https://parserator.datamade.us/usaddress/) web service which takes in an address input (e.g. `123 main st chicago il`) and parse/break it down into different address components.

### Notes

This was my first time messing around with Django and Docker (and handling API requests).  While it took much longer than the expected time to complete it, this shows that I am determined to keep going and quickly learn a new skill/technology.  However, I am very familiar with Python and JavaScript so that helped with learning Django.

A quick note about my code: my solution semi-sorts the parsed address numbers first for some reason, but it is still assigned the correct tag.  I have tried debugging and looking up the "problem", but I haven't found anything.  I hope this won't affect my eligibility.

## Testing Instructions

* You can pull this Pull request to your local environment ([see these instructions](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally))
* Assuming you have [Docker](https://docs.docker.com/get-docker/) and [Docker Compose](https://docs.docker.com/compose/install/) installed, run the following command into the terminal:
```shell
docker-compose build
``` 
* To get the app running, run this command after the build is done and open a browser window to `localhost:8000`:
```shell
docker-compose up
```
* You can test with different inputs, such as:
   * an empty input
   * repeated address components (e.g. `123 main st chicago il 123 main st`)
   * full address with punctuation (e.g. `123 Main St. Suite 100 Chicago, IL`)
* Other than the order of the table, the output should be exactly the same as the [live Parserator app](https://parserator.datamade.us/usaddress/)
* To run the test file, type the following command (you might need to type `CTRL + C` in the terminal if the app is still running):
```shell
docker-compose -f docker-compose.yml -f tests/docker-compose.yml run --rm app
```

## Thank you!

Thank you for taking the time to look through my application and I hope to hear from you soon!
